Add more window utilities

brendan-ward opened the pull request:
Closes #496 

Adds functions to `_io.pyx` to assist with calculating the data window, as well as utilities for calculating the union or intersection of windows.

Pending travis, this is ready to go.

sgillies commented:
@brendan-ward I'm :+1: but would like to surface these from some place other than `_io`. How about import them into `rasterio/__init__.py` like I did for `from rasterio._base import eval_window, window_shape, window_index`?

brendan-ward commented:
@sgillies I've done so.  However, be warned - pulling in `_io` to `rasterio/__init__.py` pulls in numpy, which may be a lot heavier than you want pulled in everywhere (I recall seeing a refactor a while back to trim back where numpy is imported, but don't recall the details).

Alternatively, I can wrap these functions in `rasterio/__init__.py` which could limit when numpy is imported:

```
def get_data_window(arr):
    from rasterio._io import get_data_window
    return get_data_window(arr)
```

sgillies commented:
@brendan-ward I forgot about the numpy import... maybe the _io module is fine for now, I'm just a little wary about these functions from _io leaking out of the project into userland.

brendan-ward commented:
@sgillies maybe wrapped functions would be fine then?  We do that in several of other places...

perrygeo commented:
@brendan-ward We should define the behavior of `intersect_windows` when the windows don't overlap. Right now, it returns an invalid window:

```
    intersect_windows([
        ((0, 6), (3, 6)),
        ((7, 12), (7, 12))]) == ((7, 6), (7, 6))
```

How can we catch this case? and what is the expected behavior or return value?

geowurster commented:
@sgillies @brendan-ward For my own knowledge, why keep the numpy related functions to modules that have already imported numpy?  Is the import operation that expensive?

sgillies commented:
@geowurster yes, importing numpy is costly. My work on isolating numpy imports was motivated by a desire to make `rio-info` faster, and it made it (IIRC) 5x faster.

geowurster commented:
@sgillies Got it - thanks.

@perrygeo @brendan-ward Seems like non-intersecting windows should raise an exception since the user thinks they intersect but in fact they do not.

brendan-ward commented:
I think I've addressed everything.  I changed function names to better match the semantics.  I added a test to determine if windows intersect; that way folks have more to rely on than just trying to intersect windows and having it raise an exception.  Trying to get the intersection of windows that don't intersect now returns a `ValueError`.

There's probably a slick `numpy` method for optimizing the test for intersects for a series of windows, but my brain simply wasn't getting there tonight.   Brute force pairwise comparison of ranges with fail-fast logic got the job done.

brendan-ward commented:
Squashed commits and ready to go.
